streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSlidingWindowAggregate.java

```diff
@@ -467,13 +459,35 @@ private boolean rightWindowIsNotEmpty(final ValueAndTimestamp<VAgg> rightWinAgg,
             return rightWinAgg != null && rightWinAgg.timestamp() > inputRecordTimestamp;
         }
 
+        @Override
+        protected void maybeForwardFinalResult(final Record<KIn, VIn> record, final long windowCloseTime) {
+            if (!shouldEmitFinal(windowCloseTime)) {
+                return;
+            }
+
+            final long emitRangeUpperBoundExclusive = windowCloseTime - windows.timeDifferenceMs();
+
+            if (emitRangeUpperBoundExclusive <= 0) {
+                // Sliding window's start and end timestamps are inclusive, so
+                // the window is not closed if emitRangeUpperBoundExclusive is 0,
+                // and we shouldn't emit
+                return;
+            }
+
+            final long emitRangeLowerBoundInclusive = lastEmitWindowCloseTime == ConsumerRecord.NO_TIMESTAMP ?
+                0L : lastEmitWindowCloseTime - windows.timeDifferenceMs();
```

Comment:
Should this be:
```
final long emitRangeLowerBoundInclusive = Math.max(0L, lastEmitWindowCloseTime - windows.timeDifferenceMs());
```

Otherwise, if `lastEmitWindowCloseTime < windows.timeDifferenceMs()` the result could still be negative?

Reply:
So far it seems `lastEmitWindowCloseTime` should always be no smaller than window size in either sliding or time windows, but when there's bugs it's possible that the read value from the processor metadata is small. I will update it accordingly.